target/linux/airoha/patches-6.12/999-en7523.patch

```diff
@@ -0,0 +1,1323 @@
+ 		msg0 |= FIELD_PREP(QDMA_ETH_TXMSG_TCO_MASK, 1) |
+ 			FIELD_PREP(QDMA_ETH_TXMSG_UCO_MASK, 1) |
+ 			FIELD_PREP(QDMA_ETH_TXMSG_ICO_MASK, 1);
+@@ -1977,7 +2042,8 @@ static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
+ 			__be16 csum = cpu_to_be16(skb_shinfo(skb)->gso_size);
+ 
+ 			tcp_hdr(skb)->check = (__force __sum16)csum;
+-			msg0 |= FIELD_PREP(QDMA_ETH_TXMSG_TSO_MASK, 1);
++			if (!airoha_is(qdma->eth, en7523))
++				msg0 |= FIELD_PREP(QDMA_ETH_TXMSG_TSO_MASK, 1);
+ 		}
+ 	}
+ 
+@@ -2031,6 +2097,11 @@ static netdev_tx_t airoha_dev_xmit(struct sk_buff *skb,
+ 		val = FIELD_PREP(QDMA_DESC_LEN_MASK, len);
+ 		if (i < nr_frags - 1)
+ 			val |= FIELD_PREP(QDMA_DESC_MORE_MASK, 1);
++		else if (len < 512)
++			val = FIELD_PREP(QDMA_DESC_LEN_MASK, len+4);
++		else
++			val = FIELD_PREP(QDMA_DESC_LEN_MASK, len+4);
++
+ 		WRITE_ONCE(desc->ctrl, cpu_to_le32(val));
+ 		WRITE_ONCE(desc->addr, cpu_to_le32(addr));
+ 		val = FIELD_PREP(QDMA_DESC_NEXT_ID_MASK, index);
+@@ -2153,7 +2224,7 @@ static int airoha_qdma_set_chan_tx_sched(struct airoha_gdm_port *port,
+ {
+ 	int i;
+ 
+-	for (i = 0; i < AIROHA_NUM_TX_RING; i++)
++	for (i = 0; i < AIROHA_NUM_TX_RING(port->soc); i++)
+ 		airoha_qdma_clear(port->qdma, REG_QUEUE_CLOSE_CFG(channel),
+ 				  TXQ_DISABLE_CHAN_QUEUE_MASK(channel, i));
+ 
+@@ -2557,7 +2628,7 @@ static int airoha_tc_htb_alloc_leaf_queue(struct airoha_gdm_port *port,
+ 	}
+ 
+ 	set_bit(channel, port->qos_sq_bmap);
+-	opt->qid = AIROHA_NUM_TX_RING + channel;
++	opt->qid = AIROHA_NUM_TX_RING(port->soc) + channel;
+ 
+ 	return 0;
+ }
+@@ -2569,7 +2640,7 @@ static int airoha_qdma_set_rx_meter(struct airoha_gdm_port *port,
+ 	struct airoha_qdma *qdma = port->qdma;
+ 	int i;
+ 
+-	for (i = 0; i < ARRAY_SIZE(qdma->q_rx); i++) {
++	for (i = 0; i < AIROHA_NUM_RX_RING(qdma->eth->soc); i++) {
+ 		int err;
+ 
+ 		if (!qdma->q_rx[i].ndesc)
+@@ -2781,7 +2852,7 @@ static int airoha_tc_get_htb_get_leaf_queue(struct airoha_gdm_port *port,
+ 		return -EINVAL;
+ 	}
+ 
+-	opt->qid = AIROHA_NUM_TX_RING + channel;
++	opt->qid = AIROHA_NUM_TX_RING(port->soc) + channel;
+ 
+ 	return 0;
+ }
+@@ -3031,8 +3102,8 @@ static int airoha_alloc_gdm_port(struct airoha_eth *eth,
+ 	}
+ 
+ 	dev = devm_alloc_etherdev_mqs(eth->dev, sizeof(*port),
+-				      AIROHA_NUM_NETDEV_TX_RINGS,
+-				      AIROHA_NUM_RX_RING);
++				      AIROHA_NUM_NETDEV_TX_RINGS(eth->soc),
++				      AIROHA_NUM_RX_RING(eth->soc));
+ 	if (!dev) {
+ 		dev_err(eth->dev, "alloc_etherdev failed\n");
+ 		return -ENOMEM;
+@@ -3043,9 +3114,9 @@ static int airoha_alloc_gdm_port(struct airoha_eth *eth,
+ 	dev->ethtool_ops = &airoha_ethtool_ops;
+ 	dev->max_mtu = AIROHA_MAX_MTU;
+ 	dev->watchdog_timeo = 5 * HZ;
+-	dev->hw_features = NETIF_F_IP_CSUM | NETIF_F_RXCSUM |
++	dev->hw_features = /* NETIF_F_IP_CSUM  |*/ NETIF_F_RXCSUM |
+ 			   NETIF_F_TSO6 | NETIF_F_IPV6_CSUM |
+-			   NETIF_F_SG | NETIF_F_TSO |
++			   NETIF_F_SG | /*NETIF_F_TSO |*/
+ 			   NETIF_F_HW_TC;
+ 	dev->features |= dev->hw_features;
```

Comment:
These should be reverted to the initial settings. But some setup might result in faster speeds so we should test some permutations.